--- ../src-base/minecraft/net/minecraft/tileentity/TileEntity.java
+++ ../src-work/minecraft/net/minecraft/tileentity/TileEntity.java
@@ -1,6 +1,8 @@
 package net.minecraft.tileentity;
 
 import javax.annotation.Nullable;
+
+import carpet.helpers.TileEntityCraftingTable;
 import net.minecraft.block.Block;
 import net.minecraft.block.BlockJukebox;
 import net.minecraft.block.state.IBlockState;
@@ -29,6 +31,9 @@
     private int field_145847_g = -1;
     protected Block field_145854_h;
 
+    //CM
+    public String cm_name() { return "Other Tile Entity"; }
+
     private static void func_190560_a(String p_190560_0_, Class <? extends TileEntity > p_190560_1_)
     {
         field_190562_f.func_82595_a(new ResourceLocation(p_190560_0_), p_190560_1_);
@@ -57,7 +62,7 @@
 
     public void func_145839_a(NBTTagCompound p_145839_1_)
     {
-        this.field_174879_c = new BlockPos(p_145839_1_.func_74762_e("x"), p_145839_1_.func_74762_e("y"), p_145839_1_.func_74762_e("z"));
+        this.field_174879_c = new BlockPos(p_145839_1_.func_74762_e("x"), p_145839_1_.func_74762_e("y"), p_145839_1_.func_74762_e("z")).wrap();
     }
 
     public NBTTagCompound func_189515_b(NBTTagCompound p_189515_1_)
@@ -258,7 +263,7 @@
 
     public void func_174878_a(BlockPos p_174878_1_)
     {
-        this.field_174879_c = p_174878_1_.func_185334_h();
+        this.field_174879_c = p_174878_1_.func_185334_h().wrap();
     }
 
     public boolean func_183000_F()
@@ -307,5 +312,6 @@
         func_190560_a("command_block", TileEntityCommandBlock.class);
         func_190560_a("shulker_box", TileEntityShulkerBox.class);
         func_190560_a("bed", TileEntityBed.class);
+        func_190560_a("crafting_table", TileEntityCraftingTable.class);
     }
 }
