--- ../src-base/minecraft/net/minecraft/util/math/BlockPos.java
+++ ../src-work/minecraft/net/minecraft/util/math/BlockPos.java
@@ -1,5 +1,6 @@
 package net.minecraft.util.math;
 
+import carpet.torus.TorusUtil;
 import com.google.common.collect.AbstractIterator;
 import com.google.common.collect.Lists;
 import java.util.Iterator;
@@ -166,6 +167,10 @@
         return ((long)this.func_177958_n() & field_177994_h) << field_177988_g | ((long)this.func_177956_o() & field_177995_i) << field_177987_f | ((long)this.func_177952_p() & field_177993_j) << 0;
     }
 
+    public BlockPos wrap(){
+        return new BlockPos(func_177958_n() & TorusUtil.X_MASK, func_177956_o(), func_177952_p()&TorusUtil.Z_MASK);
+    }
+
     public static BlockPos func_177969_a(long p_177969_0_)
     {
         int i = (int)(p_177969_0_ << 64 - field_177988_g - field_177990_b >> 64 - field_177990_b);
